### ISSUE TITLE
[DTensor] aten.max.dim returns wrong indices when using DTensor

### ISSUE THREAD

**Issue by qqq6op** [pytorch/pytorch#168940]
### 🐛 Describe the bug

I found that current strategy of `aten.max.dim` may get incorrect indices output if sharded the dim for maximization.

Sample code:
```python
import torch
from torch.distributed.tensor import distribute_tensor, Shard

from torch.testing._internal.common_utils import run_tests
from torch.testing._internal.distributed._tensor.common_dtensor import DTensorTestBase, with_comms


class TestRegisterSharding(DTensorTestBase):
    @with_comms
    def test_max_dim(self):
        mesh = self.build_device_mesh()

        x = torch.randn(4, 4, device="cuda")

        max_value, max_indices = torch.max(x, dim=1)

        dist_x = distribute_tensor(x, mesh, [Shard(1)])

        dist_max_value, dist_max_indices = torch.max(dist_x, dim=1)

        print("x:", x)
        print("max_value:", max_value)
        print("max_indices:", max_indices)
        print("dist_max_value:", dist_max_value.full_tensor())
        print("dist_max_indices:", dist_max_indices.full_tensor())


if __name__ == "__main__":
    run_tests()
```

Result:
```python
x: tensor([[-1.6165,  0.5685, -0.5102, -0.9113],
        [-1.1555, -0.2262, -1.2891,  1.0654],
        [-0.7167, -0.5333,  0.2078, -0.9798],
        [ 0.7447, -0.2395,  0.2737,  0.0920]], device='cuda:0')
max_value: tensor([0.5685, 1.0654, 0.2078, 0.7447], device='cuda:0')
max_indices: tensor([1, 3, 2, 0], device='cuda:0')
dist_max_value: tensor([0.5685, 1.0654, 0.2078, 0.7447], device='cuda:0')
dist_max_indices: tensor([0, 0, 0, 0], device='cuda:0')
```

Each rank gets a shape(4, 1) local tensor to call `max.dim` in this case, and the local result of max indices is [0, 0, 0, 0]. The framework doesn't process the offset of index, which leads to an incorrect global result when the relavant dim is sharded.

Is there a good way to implement a strategy that supports sharding the index dim?

### Versions

torch v2.9.0

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @tianyu-l @XilunWu @SherlockNoMad

**Comment by skpark-rh** [comment]
I just submitted a [PR](https://github.com/pytorch/pytorch/pull/168983) that implements the indices in a global manner. My [PR](https://github.com/pytorch/pytorch/pull/168983) won't directly fix this issue as this seems to be a problem with the max function but you may be able to leverage the argmax function to do this. I wonder if I can leverage what I wrote to fix this... I'll take a quick look.

**Comment by skpark-rh** [comment]
So aten.max.dim is not a linear reduction because of the indices returned. The only solution I see is implementing a custom handler to deal with the offsets or use a strategy with linear_reduction turned off and replicate the sharded dimension.  Since you want the indices sharded I'd say the first option is the one you are looking for. I can get a implementation going to accommodate the max/min with the dim provided.

**Comment by qqq6op** [comment]
> So aten.max.dim is not a linear reduction because of the indices returned. The only solution I see is implementing a custom handler to deal with the offsets or use a strategy with linear_reduction turned off and replicate the sharded dimension. Since you want the indices sharded I'd say the first option is the one you are looking for. I can get a implementation going to accommodate the max/min with the dim provided.

Thanks a lot! Since the common pipeline of DTensor is not aware of when and which parameters require offset handling, defining a custom handler for specific aten ops is currently the only solution. The implementation in your [PR](https://github.com/pytorch/pytorch/pull/168983) is a good reference for me.

**Comment by skpark-rh** [comment]
Yes. So I dug deeper into this and the mismatch for my fix is happening because a thread prematurely go to the next sample while the other 3 stays on the same sample.  I added a barrier to ensure that every thread gets to the same point but to no avail... The async wait doesn't seem to work. I know if the threads are held in the same place then this problem goes away. I can submit a draft with my branch so that you can clone it to verify.
